demo-circuit/src/naive_threshold_sig_w_key_rotation/tests.rs

```diff
@@ -0,0 +1,1180 @@
+        test_naive_threshold_circuit(circuit, debug_only, max_pks, num_custom_fields, None);
+
+    println!("Failing constraint: {:?}", failing_constraint);
+    assert!(failing_constraint
+        .clone()
+        .unwrap()
+        .contains("require(sc_wcert_hash == H(wcert)/conditional_equals"));
+}
+
+#[serial]
+#[test]
+fn test_naive_threshold_sig_w_key_rotation_circuit_wrong_prev_cert_data_hash() {
+    let debug_only = true;
+
+    let (
+        max_pks,
+        num_custom_fields,
+        valid_sigs,
+        threshold,
+        is_first_certificate,
+        expected_threshold_results,
+        change_signing_keys,
+        change_master_keys,
+    ) = generate_circuit_test_data(TestDataType::AllRandomNextCert);
+    assert!(is_first_certificate == false);
+    assert!(expected_threshold_results == true);
+
+    println!("Checking wrong prev_cert_data_hash when is_first_certificate is false");
+    let mut circuit = get_test_circuit_instance(
+        max_pks,
+        num_custom_fields,
+        valid_sigs,
+        threshold,
+        is_first_certificate,
+        change_signing_keys,
+        change_master_keys,
+    );
+
+    let rng = &mut thread_rng();
+
+    circuit.prev_cert_data_hash = rng.gen();
+
+    let failing_constraint =
+        test_naive_threshold_circuit(circuit, debug_only, max_pks, num_custom_fields, None);
+
+    println!("Failing constraint: {:?}", failing_constraint);
+    assert!(failing_constraint
+        .clone()
+        .unwrap()
+        .contains("require(sc_prev_wcert_hash == H(prev_wcert)/conditional_equals"));
+}
+
+#[serial]
+#[test]
+fn test_naive_threshold_sig_w_key_rotation_circuit_wrong_prev_scb_validators_keys_root() {
+    let debug_only = true;
+
+    let (
+        max_pks,
+        num_custom_fields,
+        valid_sigs,
+        threshold,
+        is_first_certificate,
+        expected_threshold_results,
+        change_signing_keys,
+        change_master_keys,
+    ) = generate_circuit_test_data(TestDataType::AllRandomNextCert);
+    assert!(is_first_certificate == false);
+    assert!(expected_threshold_results == true);
+
+    println!("Checking wrong prev_scb_validators_keys_root when is_first_certificate is false");
+    let mut circuit = get_test_circuit_instance(
+        max_pks,
+        num_custom_fields,
+        valid_sigs,
+        threshold,
+        is_first_certificate,
+        change_signing_keys,
+        change_master_keys,
+    );
+
+    let rng = &mut thread_rng();
+
+    //prev_scb_validators_keys_root is in first prev custom fields
+    circuit.prev_withdrawal_certificate.custom_fields[0] = rng.gen();
+
+    let failing_constraint =
+        test_naive_threshold_circuit(circuit, debug_only, max_pks, num_custom_fields, None);
+
+    println!("Failing constraint: {:?}", failing_constraint);
+    assert!(failing_constraint
+        .clone()
+        .unwrap()
+        .contains("require(sc_prev_wcert_hash == H(prev_wcert)/conditional_equals"));
```

Comment:
Unfortunately we don't get the expected failing constraint as the check on the hash comes before the one we are testing. Perhaps we can move the checks on the public inputs at the very end of the circuit ?